[ADR] Custom operation in new Sylius API (#11478)
Sylius/Sylius: adr/2020_05_13_handling_non_crud_operations_in_api.md

# Handling Non-CRUD operations in API

* Status: proposed
* Date: 2020-05-13

## Context and Problem Statement

> Handling Non-CRUD operation over REST API is not trivial. Once operations are beyond Create, Read, Update, Delete, 

Reviewer: ```suggestion
Handling non-CRUD operation over REST API is not trivial. Once operations are beyond Create, Read, Update, Delete, 
```

Reviewer: Btw. what about rephrasing this sentence to something more like:

> Once operations are beyond Create (`POST`/`PUT`), Read (`GET`), ..., there is no clear recommendation...

which can be mapped directly to `POST`/`PUT`, `GET`, `PUT`/`PATCH`, `DELETE` HTTP methods there is no clear recommendation 
how to map others actions. These actions include, but are not limited to, changes of states (described in the form of 
workflow in a state machine) or command execution.

## Decision Drivers

> * Solution should not be limited by its convention. We may need to support two different actions of the same name, 

Reviewer: > We may need to support two different actions of the same name, but with different business logic.

I'm not sure whether I understand it correctly. Does it mean these two actions correspond to two different transitions leading to the same state?

> but with different business logic. E.g., One object may have two transitions with the same name but described by 

Reviewer: ```suggestion
but with different business logic. E.g., one object may have two transitions with the same name but described by 
```

two different state machines.
* Solution should allow for an easy understanding of what is expected behavior of its execution.

> * Solution should provide easy to implement a way of exposing the next possible actions. (according to HATEOAS paradigm).

Reviewer: ```suggestion
* Solution should provide easy to implement a way of exposing the next possible actions, according to HATEOAS paradigm.
```

* Solution should provide a clean way of adding additional fields during the execution of requested operations.

## Considered Options

> All options are presented on the example of making a ship's operations over the shipment, that is ready to be shipped.

Reviewer: ```suggestion
All options are presented on the example of making a `ship` operation over the shipment, that is ready to be shipped.
```

> Expectations: be able to call `ship` transition on the `Shipment` with the XYZ tracking code. 

Reviewer: ```suggestion
Expectations: being able to call `ship` transition on the `Shipment` with the XYZ tracking code. 
```

### Enrich resource data with possible operations

It is possible to add additional fields or reuse current ones to accept operations and leverage HTTP Verbs. In 
this solution we can use regular `Update` request to execute additional logic:

PUT `/api/orders/1/shipment/1`

```json
{
  "status": "ship",

>   "tacking_code": "XYZ"

Reviewer 2: ```suggestion
  "tracking_code": "XYZ"
```
not only here

}
```

> This solution works for all standards supported by ApiPlatform(JSON-LD, Json HAL, GraphQL)

Reviewer: ```suggestion
This solution works for all standards supported by ApiPlatform(JSON-LD, JSON HAL, GraphQL).
```

> * Good, because it does not introduce any new `verbs` (which may be considered as a bad practice for REST API)

Reviewer: Btw. I'm not sure whether we should use "verbs" or "HTTP methods"/"HTTP request methods".

> * Good, because it does not introduce any new `endpoints`.

Reviewer: ```suggestion
* Good, because it does not introduce any new endpoints.
```

What's the reason behind treating "endpoint" as a code?

`"status": "ship"` results with `"status": "shipped"`).
* Bad, because it is not clear how to expose possible transitions. Should a separate endpoint be created, or data should
 be further enriched with the `possibleTransitions` field? It is not clear how non-state machine operations should be 
 handled as well. 
* Bad, because it uses `PUT` operation for action, which is nonidempotent, which may mislead end clients.
* Bad, because it allows for using only one type of transition per field.

### Defining custom operations in the style of the command pattern

We may trace operations as separate resources. We may operate with these operations using regular HTTP Verbs. All these 
operations should be named as nouns. 

POST `/api/orders/1/shipment/1/ship-attempts`

```json
{
  "tracking_code": "XYZ"
}
```

* Good, because it does not introduce any new `verbs` (which may be considered as a bad practice for REST API)
* Good, because there may be a straight forward way to expose possible transitions (GET `/api/orders/1/shipment/1/ship-attempts`)
* Good, because it provides a clear solution for all custom endpoints. 

> * Good, because it uses `POST` operation for action, which is nonidempotent. 

Reviewer: ```suggestion
* Good, because it uses `POST` operation for action, which is not idempotent. 
```

Makes it a bit clearer IMO.

* Bad, because state changes are not traceable resources in Sylius. They do not have unique identifiers to check their 
status, nor are they possible for retries.

> * Bad, because there is no clear solution on how to declare links in Json-LD format.

Reviewer: ```suggestion
* Bad, because there is no clear solution on how to declare links in JSON-LD format.
```

### Taking advantage of the `Controller` REST archetype

> This approach was described in the REST API Design Rulebook, which states 4 Archetypes: `Document`, `Collection`, `Store`,

Reviewer: Could you link the rulebook in the references section?

> This approach was described in the REST API Design Rulebook, which states 4 Archetypes: `Document`, `Collection`, `Store`,

Reviewer: ```suggestion
This approach was described in the REST API Design Rulebook, which states 4 archetypes: `Document`, `Collection`, `Store`,
```

>  and `Controller`.`Controller` archetype allows for custom operations over the resource with the custom verb as the last 

Reviewer: ```suggestion
and `Controller`. `Controller` archetype allows for custom operations over the resource with the custom verb as the last 
```

>  part of the URL. However, these custom actions cannot duplicate the logic of any HTTP Verbs.

Reviewer: ```suggestion
part of the URL. However, these custom actions cannot duplicate the logic of any HTTP verbs.
```

POST `/api/orders/1/shipment/1/ship`

```json
{
  "tracking_code": "XYZ"
}
```

* Good, because it is already supported by Sylius Admin REST API - which reduces the number of possible problems during 
migration.
* Good, because it is straight-forward and easy to comprehend.
* Good, because it provides a clear solution for all custom solutions. 

> * Good, because it uses `POST` operation for action, which is nonidempotent. 

Reviewer: ```suggestion
* Good, because it uses `POST` operation for action, which is not idempotent. 
```

> * Bad, because there is no clear solution on how to declare links in Json-LD format.

Reviewer: ```suggestion
* Bad, because there is no clear solution on how to declare links in JSON-LD format.
```

* Bad, because usage of verbs in URL may be considered as a bad practice. 

### Google recommendation for custom operations

Google is taking an approach of suffixing its custom endpoints with `:operation`.

POST `/api/orders/1/shipment/1:ship`

```json
{
  "tracking_code": "XYZ"
}
```

* Good, because it is straight-forward and easy to comprehend.
* Good, because it provides a clear solution for all custom solutions. 
* Good, because it uses `POST` operation for action, which is nonidempotent. 
* Bad, because it seems to be unnatural for standard REST endpoints.

> * Bad, because it can be hard to achieve with the Api Platform.

Reviewer: ```suggestion
* Bad, because it can be hard to achieve with the API Platform.
```
